File: server/routers/websocket.py
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Query

from infra.log import logger
import services.state as state

router = APIRouter()


@router.websocket("/bridge/bot")
async def ws_bot(
    ws: WebSocket,
    token: str = Query(default=""),
):
    bot_token = token or (ws.headers.get("x-astron-bot-token", ""))
    if not await state.token_manager.validate(bot_token):
        await ws.accept()
        await ws.close(code=4001, reason="Invalid or missing bot token")
        logger.warning("Bot connection rejected: invalid token {}...", bot_token[:10])
        return

    await ws.accept()

    if not await state.bridge.register_bot(bot_token, ws):
        await ws.send_json({"error": "Another bot is already connected with this token"})
        await ws.close(code=4002, reason="Bot already connected")
        logger.warning("Bot connection rejected: duplicate token {}...", bot_token[:10])
        return

    logger.info("Bot connected: {}...", bot_token[:10])
    await state.bridge.notify_bot_connected(bot_token)
    try:
        while True:
            raw = await ws.receive_text()
            await state.bridge.handle_bot_message(bot_token, raw)
    except WebSocketDisconnect:
        logger.info("Bot disconnected: {}...", bot_token[:10])
    except Exception:
        logger.exception("Bot connection error: {}...", bot_token[:10])
    finally:
        await state.bridge.unregister_bot(bot_token)
        await state.bridge.notify_bot_disconnected(bot_token)

File: backend/internal/router/websocket.go
package router

import (
	"context"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
	"github.com/rs/zerolog/log"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"

	"astron-claw/backend/internal/model"
	"astron-claw/backend/internal/service"
)

var wsTracer = otel.Tracer("astron-claw/router/websocket")

var wsUpgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return true },
}

func (app *App) wsBot(c *gin.Context) {
	botToken := c.Query("token")
	if botToken == "" {
		botToken = c.GetHeader("x-astron-bot-token")
	}

	if !app.TokenMgr.Validate(c.Request.Context(), botToken) {
		// Accept then close with error code
		conn, err := wsUpgrader.Upgrade(c.Writer, c.Request, nil)
		if err != nil {
			log.Error().Err(err).Msg("WS upgrade failed for invalid token")
			return
		}
		msg := websocket.FormatCloseMessage(model.ErrWSInvalidToken.HTTPStatus, model.ErrWSInvalidToken.Message)
		_ = conn.WriteMessage(websocket.CloseMessage, msg)
		conn.Close()
		log.Warn().Str("token", botToken).Msg("Bot connection rejected: invalid token")
		return
	}

	conn, err := wsUpgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		log.Error().Err(err).Msg("WS upgrade failed")
		return
	}

	clientAddr := c.ClientIP()
	botConn := &service.BotConn{
		Conn:  conn,
		Token: botToken,
	}

	ctx, regSpan := wsTracer.Start(c.Request.Context(), "bot.connection.register",
		trace.WithSpanKind(trace.SpanKindInternal))
	regSpan.SetAttributes(
		attribute.String("astron.token_id", botToken),
		attribute.String("astron.client_addr", clientAddr),
	)

	if err := app.Bridge.RegisterBot(ctx, botToken, botConn); err != nil {
		regSpan.End()
		log.Error().Err(err).Str("token", botToken).Msg("Failed to register bot")
		conn.Close()
		return
	}
	regSpan.End()

	log.Info().Str("token", botToken).Str("from", clientAddr).Msg("Bot connected")
	app.Bridge.NotifyBotConnected(botToken)

	defer func() {
		unregCtx, unregSpan := wsTracer.Start(context.Background(), "bot.connection.unregister",
			trace.WithSpanKind(trace.SpanKindInternal))
		unregSpan.SetAttributes(
			attribute.String("astron.token_id", botToken),
		)
		cleanupCtx, cancel := context.WithTimeout(unregCtx, 10*time.Second)
		defer cancel()
		app.Bridge.UnregisterBot(cleanupCtx, botToken, botConn)
		unregSpan.End()
	}()

	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				log.Info().Str("token", botToken).Str("from", clientAddr).Msg("Bot disconnected normally")
			} else if websocket.IsUnexpectedCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				log.Info().Str("token", botToken).Str("from", clientAddr).Err(err).Msg("Bot disconnected unexpectedly")
			} else {
				log.Error().Err(err).Str("token", botToken).Msg("Bot connection error")
			}
			return
		}

		app.Bridge.HandleBotMessage(context.Background(), botToken, string(message))
	}
}
